File: android/app/src/main/java/com/example/audiomodem/modem/Constellation.kt
package com.example.audiomodem.modem

import kotlin.math.sqrt

/**
 * QAM constellation mapping and demapping.
 */
enum class Modulation(val bitsPerSymbol: Int) {
    QPSK(2),
    QAM16(4),
    QAM64(6);

    override fun toString(): String = when (this) {
        QPSK -> "QPSK"
        QAM16 -> "16-QAM"
        QAM64 -> "64-QAM"
    }
}

class Constellation(val mod: Modulation) {
    // Points stored as interleaved [re0, im0, re1, im1, ...]
    private val pointsRe: DoubleArray
    private val pointsIm: DoubleArray

    init {
        val (re, im) = when (mod) {
            Modulation.QPSK -> generateQPSK()
            Modulation.QAM16 -> generateQAM(4)
            Modulation.QAM64 -> generateQAM(8)
        }
        // Normalize to unit average power
        var avgPower = 0.0
        for (i in re.indices) {
            avgPower += re[i] * re[i] + im[i] * im[i]
        }
        avgPower /= re.size
        val scale = 1.0 / sqrt(avgPower)
        for (i in re.indices) {
            re[i] *= scale
            im[i] *= scale
        }
        pointsRe = re
        pointsIm = im
    }

    private fun generateQPSK(): Pair<DoubleArray, DoubleArray> {
        val re = doubleArrayOf(1.0, -1.0, -1.0, 1.0)
        val im = doubleArrayOf(1.0, 1.0, -1.0, -1.0)
        return Pair(re, im)
    }

    private fun generateQAM(order: Int): Pair<DoubleArray, DoubleArray> {
        val size = order * order
        val re = DoubleArray(size)
        val im = DoubleArray(size)

        for (i in 0 until size) {
            val row = i / order
            val col = i % order
            val grayRow = row xor (row shr 1)
            val grayCol = col xor (col shr 1)
            re[i] = (2.0 * grayCol - order + 1)
            im[i] = (2.0 * grayRow - order + 1)
        }
        return Pair(re, im)
    }

    fun map(bits: ByteArray): Pair<Double, Double> {
        val idx = bitsToIndex(bits)
        return Pair(pointsRe[idx.coerceIn(0, pointsRe.size - 1)],
                    pointsIm[idx.coerceIn(0, pointsIm.size - 1)])
    }

    fun demap(re: Double, im: Double): ByteArray {
        var minDist = Double.MAX_VALUE
        var minIdx = 0
        for (i in pointsRe.indices) {
            val dr = re - pointsRe[i]
            val di = im - pointsIm[i]
            val d = dr * dr + di * di
            if (d < minDist) {
                minDist = d
                minIdx = i
            }
        }
        return indexToBits(minIdx, mod.bitsPerSymbol)
    }

    fun mapBits(bits: ByteArray): Pair<DoubleArray, DoubleArray> {
        val bps = mod.bitsPerSymbol
        val numSymbols = bits.size / bps
        val re = DoubleArray(numSymbols)
        val im = DoubleArray(numSymbols)

        for (i in 0 until numSymbols) {
            val subBits = bits.copyOfRange(i * bps, (i + 1) * bps)
            val (r, ii) = map(subBits)
            re[i] = r
            im[i] = ii
        }
        return Pair(re, im)
    }

    fun demapSymbols(re: DoubleArray, im: DoubleArray): ByteArray {
        val bps = mod.bitsPerSymbol
        val bits = ByteArray(re.size * bps)
        for (i in re.indices) {
            val b = demap(re[i], im[i])
            b.copyInto(bits, i * bps)
        }
        return bits
    }

    companion object {
        fun bitsToIndex(bits: ByteArray): Int {
            var idx = 0
            for (b in bits) {
                idx = (idx shl 1) or (b.toInt() and 1)
            }
            return idx
        }

        fun indexToBits(idx: Int, numBits: Int): ByteArray {
            val bits = ByteArray(numBits)
            var v = idx
            for (i in numBits - 1 downTo 0) {
                bits[i] = (v and 1).toByte()
                v = v shr 1
            }
            return bits
        }
    }
}

File: pc/internal/modem/constellation.go
package modem

import (
	"math"
)

// Modulation represents a QAM modulation scheme.
type Modulation int

const (
	ModQPSK  Modulation = 2 // 2 bits per symbol
	Mod16QAM Modulation = 4 // 4 bits per symbol
	Mod64QAM Modulation = 6 // 6 bits per symbol
)

// BitsPerSymbol returns the number of bits per constellation symbol.
func (m Modulation) BitsPerSymbol() int {
	return int(m)
}

// String returns the modulation name.
func (m Modulation) String() string {
	switch m {
	case ModQPSK:
		return "QPSK"
	case Mod16QAM:
		return "16-QAM"
	case Mod64QAM:
		return "64-QAM"
	default:
		return "Unknown"
	}
}

// Constellation holds QAM constellation points.
type Constellation struct {
	Mod    Modulation
	points []complex128
	scale  float64 // normalization factor for unit average power
}

// NewConstellation creates a new constellation for the given modulation.
func NewConstellation(mod Modulation) *Constellation {
	c := &Constellation{Mod: mod}
	switch mod {
	case ModQPSK:
		c.generateQPSK()
	case Mod16QAM:
		c.generateQAM(4) // 4x4
	case Mod64QAM:
		c.generateQAM(8) // 8x8
	default:
		c.generateQPSK()
	}
	c.normalize()
	return c
}

func (c *Constellation) generateQPSK() {
	// Gray-coded QPSK: 00, 01, 11, 10
	c.points = []complex128{
		complex(1, 1),   // 00
		complex(-1, 1),  // 01
		complex(-1, -1), // 11
		complex(1, -1),  // 10
	}
}

func (c *Constellation) generateQAM(order int) {
	// Generate square QAM constellation with Gray coding
	size := order * order
	c.points = make([]complex128, size)

	for i := 0; i < size; i++ {
		row := i / order
		col := i % order

		// Gray code mapping
		grayRow := row ^ (row >> 1)
		grayCol := col ^ (col >> 1)

		// Map to symmetric constellation
		x := float64(2*grayCol-order+1) // odd values: -3, -1, 1, 3 for 4-QAM
		y := float64(2*grayRow-order+1)

		c.points[i] = complex(x, y)
	}
}

func (c *Constellation) normalize() {
	// Calculate average power
	var avgPower float64
	for _, p := range c.points {
		avgPower += real(p)*real(p) + imag(p)*imag(p)
	}
	avgPower /= float64(len(c.points))

	// Normalize to unit average power
	c.scale = 1.0 / math.Sqrt(avgPower)
	for i := range c.points {
		c.points[i] = complex(real(c.points[i])*c.scale, imag(c.points[i])*c.scale)
	}
}

// Map maps bits to a constellation point.
func (c *Constellation) Map(bits []byte) complex128 {
	idx := bitsToIndex(bits)
	if idx >= len(c.points) {
		idx = 0
	}
	return c.points[idx]
}

// Demap finds the closest constellation point and returns the bits.
func (c *Constellation) Demap(symbol complex128) []byte {
	minDist := math.MaxFloat64
	minIdx := 0

	for i, p := range c.points {
		d := real(symbol-p)*real(symbol-p) + imag(symbol-p)*imag(symbol-p)
		if d < minDist {
			minDist = d
			minIdx = i
		}
	}

	return indexToBits(minIdx, c.Mod.BitsPerSymbol())
}

// MapBits maps a bit slice to constellation symbols.
// bits are packed as bytes (0 or 1 each).
func (c *Constellation) MapBits(bits []byte) []complex128 {
	bps := c.Mod.BitsPerSymbol()
	numSymbols := len(bits) / bps
	symbols := make([]complex128, numSymbols)

	for i := 0; i < numSymbols; i++ {
		symbols[i] = c.Map(bits[i*bps : (i+1)*bps])
	}
	return symbols
}

// DemapSymbols demaps constellation symbols back to bits.
func (c *Constellation) DemapSymbols(symbols []complex128) []byte {
	bps := c.Mod.BitsPerSymbol()
	bits := make([]byte, 0, len(symbols)*bps)

	for _, s := range symbols {
		bits = append(bits, c.Demap(s)...)
	}
	return bits
}

func bitsToIndex(bits []byte) int {
	idx := 0
	for _, b := range bits {
		idx = (idx << 1) | int(b&1)
	}
	return idx
}

func indexToBits(idx, numBits int) []byte {
	bits := make([]byte, numBits)
	for i := numBits - 1; i >= 0; i-- {
		bits[i] = byte(idx & 1)
		idx >>= 1
	}
	return bits
}
